### PR TITLE
[GCS Actor Management] Gcs actor management broken detached actor

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This addressed multiple issues.

1. If timeout is triggered by the promise, and a GCS response comes after that, it segfaults because promise is allocated on the stack.
=> Fix it by wrapping it in a shared pointer. (so that the promise is not deallocated)

2. We are holding gil when we are calling `GetNamedActorHandle`. This leads us to hang the function because we are calling `get_py_stack` method which needs to hold gil. This causes a deadlock. 
=> Nogil when this method is called.

3. Since `GetNamedActorHandle` can have multiple `Status` (timeout, not found), we should return them from the method. We return the pair to be consistent with the `GetActorHandle` method

4. Currently, the GCS_ACTOR_SERVICE_ENABELD wasn't actually turned on the CI because of the bazel issue I wasn't aware of. This PR will turn it on. 

## Related issue number

https://github.com/ray-project/ray/issues/9383

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -1081,11 +1078,15 @@ cdef class CoreWorker:
     def get_named_actor_handle(self, const c_string &name):
         cdef:
             # NOTE: This handle should not be stored anywhere.
```
Comment (stephanie-wang) [inline python/ray/_raylet.pyx:1080]: Move the note down one line?

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -1321,9 +1322,10 @@ const ActorHandle *CoreWorker::GetNamedActorHandle(const std::string &name) {
   // There should be no risk of deadlock because we don't hold any
   // locks during the call and the RPCs run on a separate thread.
   ActorID actor_id;
-  auto ready_promise = std::promise<void>();
+  std::shared_ptr<std::promise<void>> ready_promise =
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc:1325]: Nice catch!

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -1081,11 +1078,15 @@ cdef class CoreWorker:
...
-            const CActorHandle* c_actor_handle = (
+            pair[const CActorHandle*, CRayStatus] named_actor_handle_pair
+            const CActorHandle* c_actor_handle
+
+        with nogil:
```
Comment (stephanie-wang) [inline python/ray/_raylet.pyx:1084]: Can you add a comment explaining why nogil?

**Review by stephanie-wang — APPROVED** [review]

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx]: Good catch! Just addressed

File: python/ray/_raylet.pyx (hunk 3 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:1086]: Added!

**Review by rkooo567 — COMMENTED** [review]
All comments are addressed!

**Comment by rkooo567** [comment]
@mehrdadn Can you check if it is the right way to set the env var?

**Review by ffbin — APPROVED** [review]

**Comment by mehrdadn** [comment]
@rkooo567 Yup it looks fine to me. (Though it's not _setting_ the env var, but merely propagating it.)

**Comment by rkooo567** [comment]
@mehrdadn Yes! That's the behavior I want. I want to propagate this env variable so that only the 4th CI turns it on. Thank you for checking it out :)

**Comment by rkooo567** [comment]
@stephanie-wang Can you merge this PR?

**Pull request merged**